chain/client/src/client.rs

```diff
@@ -703,29 +678,31 @@ impl Client {
 
         if tip.last_block_hash != self.doomslug.get_tip().0 {
             // We need to update the doomslug tip
-            let last_ds_final_hash =
-                self.chain.get_block_header(&tip.last_block_hash)?.inner_rest.last_ds_final_block;
-            let last_ds_final_height = if last_ds_final_hash == CryptoHash::default() {
-                0
+            let last_final_hash =
+                self.chain.get_block_header(&tip.last_block_hash)?.inner_rest.last_final_block;
+            let last_final_height = if last_final_hash == CryptoHash::default() {
+                self.chain.genesis().inner_lite.height
             } else {
-                self.chain.get_block_header(&last_ds_final_hash)?.inner_lite.height
+                self.chain.get_block_header(&last_final_hash)?.inner_lite.height
             };
 
             self.doomslug.set_tip(
                 Instant::now(),
                 tip.last_block_hash,
-                self.chain.get_my_approval_reference_hash(tip.last_block_hash),
                 tip.height,
-                last_ds_final_height,
+                last_final_height,
             );
         }
 
         Ok(())
     }
 
-    pub fn send_approval(&mut self, approval: Approval) -> Result<(), Error> {
-        let parent_hash = approval.parent_hash;
-        let next_epoch_id = self.runtime_adapter.get_epoch_id_from_prev_block(&parent_hash)?;
+    pub fn send_approval(
+        &mut self,
+        parent_hash: &CryptoHash,
+        approval: Approval,
+    ) -> Result<(), Error> {
+        let next_epoch_id = self.runtime_adapter.get_epoch_id_from_prev_block(parent_hash)?;
```

Comment:
is it guaranteed that `parent_hash` is the hash of the block that the approval points to?

Reply:
Yes